core/src/mindustry/core/UI.java

```diff
@@ -579,22 +609,64 @@ public static String formatTime(float ticks){
     }
 
     public static String formatAmount(long number){
+        return formatAmount(number, 1);
+    }
+
+    public static String formatAmount(long number, int decimalPlaces){
+        return formatAmount(number, decimalPlaces, false);
+    }
+
+    public static String formatAmount(long number, boolean si){
+        return formatAmount(number, si ? 0 : 1, si);
+    }
+
+    public static String formatAmount(long number, int decimalPlaces, boolean si){
         //prevent overflow
         if(number == Long.MIN_VALUE) number ++;
 
         long mag = Math.abs(number);
         String sign = number < 0 ? "-" : "";
-        if(mag >= 1_000_000_000){
-            return sign + Strings.fixed(mag / 1_000_000_000f, 1) + "[gray]" + billions+ "[]";
-        }else if(mag >= 1_000_000){
-            return sign + Strings.fixed(mag / 1_000_000f, 1) + "[gray]" +millions + "[]";
-        }else if(mag >= 10_000){
-            return number / 1000 + "[gray]" + thousands + "[]";
-        }else if(mag >= 1000){
-            return sign + Strings.fixed(mag / 1000f, 1) + "[gray]" + thousands + "[]";
-        }else{
-            return number + "";
+
+        int unit = 0;
+        String unitString = null;
+        long value = number;
+        if(value >= 10_000_000_000_000_000L){
+            unit += 16;
+            value /= 10_000_000_000_000_000L;
+        }
+        if(value >= 100_000_000L){
+            unit += 8;
+            value /= 100_000_000L;
+        }
+        if(value >= 10_000L){
+            unit += 4;
+            value /= 10_000L;
+        }
+        if(value >= 100L){
+            unit += 2;
+            value /= 100L;
+        }
+        if(value >= 10L){
+            unit += 1;
+        }
+        while(unit > 0){
+            if(si){
+                if(numericSIUnits.containsKey(unit)){
```

Comment:
- Move the `while` into the `if`.
- Array`[]!=null` should be faster.

Reply:
> * Move the `while` into the `if`.

Can you explain me the reason?

Reply:
I was thinking of the performance at first, but it may be insignificant. And the logic order: "I'm looping finding units, seeing if I want SI for each." vs "Determine to use SI or not, then loop to find the units."

Reply:
The `unit > 19` check while you changed to arrays.